File: Services/order-service/src/main/java/com/rumal/order_service/entity/Order.java
package com.rumal.order_service.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "orders", indexes = {
        @Index(name = "idx_orders_customer_id", columnList = "customer_id"),
        @Index(name = "idx_orders_status", columnList = "status"),
        @Index(name = "idx_orders_created_at", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Order {

    @Id
    private UUID id;

    @Version
    private Long version;

    @Column(nullable = false)
    private UUID customerId;

    @Column(nullable = false, length = 120)
    private String item;

    @Column(nullable = false)
    private int quantity;

    @Column(name = "item_count", nullable = false)
    private int itemCount;

    @Column(name = "order_total", nullable = false, precision = 19, scale = 2)
    private BigDecimal orderTotal;

    @Builder.Default
    @Column(name = "currency", nullable = false, length = 3)
    private String currency = "USD";

    @Column(name = "subtotal_amount", nullable = false, precision = 19, scale = 2)
    private BigDecimal subtotal;

    @Column(name = "line_discount_total", nullable = false, precision = 19, scale = 2)
    private BigDecimal lineDiscountTotal;

    @Column(name = "cart_discount_total", nullable = false, precision = 19, scale = 2)
    private BigDecimal cartDiscountTotal;

    @Column(name = "shipping_amount", nullable = false, precision = 19, scale = 2)
    private BigDecimal shippingAmount;

    @Column(name = "shipping_discount_total", nullable = false, precision = 19, scale = 2)
    private BigDecimal shippingDiscountTotal;

    @Column(name = "total_discount", nullable = false, precision = 19, scale = 2)
    private BigDecimal totalDiscount;

    @Column(name = "coupon_code", length = 64)
    private String couponCode;

    @Column(name = "coupon_reservation_id")
    private UUID couponReservationId;

    @Column(name = "customer_note", length = 500)
    private String customerNote;

    @Column(name = "admin_note", length = 1000)
    private String adminNote;

    @Column(name = "payment_id", length = 120)
    private String paymentId;

    @Column(name = "payment_method", length = 50)
    private String paymentMethod;

    @Column(name = "payment_gateway_ref", length = 200)
    private String paymentGatewayRef;

    @Column(name = "paid_at")
    private Instant paidAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private OrderStatus status;

    @Column(name = "expires_at")
    private Instant expiresAt;

    @Column(name = "refund_amount", precision = 19, scale = 2)
    private BigDecimal refundAmount;

    @Column(name = "refund_reason", length = 500)
    private String refundReason;

    @Column(name = "refund_initiated_at")
    private Instant refundInitiatedAt;

    @Column(name = "refund_completed_at")
    private Instant refundCompletedAt;

    @Column(name = "shipping_address_id", nullable = false)
    private UUID shippingAddressId;

    @Column(name = "billing_address_id", nullable = false)
    private UUID billingAddressId;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "label", column = @Column(name = "shipping_label", length = 50)),
            @AttributeOverride(name = "recipientName", column = @Column(name = "shipping_recipient_name", nullable = false, length = 120)),
            @AttributeOverride(name = "phone", column = @Column(name = "shipping_phone", nullable = false, length = 32)),
            @AttributeOverride(name = "line1", column = @Column(name = "shipping_line_1", nullable = false, length = 180)),
            @AttributeOverride(name = "line2", column = @Column(name = "shipping_line_2", length = 180)),
            @AttributeOverride(name = "city", column = @Column(name = "shipping_city", nullable = false, length = 80)),
            @AttributeOverride(name = "state", column = @Column(name = "shipping_state", nullable = false, length = 80)),
            @AttributeOverride(name = "postalCode", column = @Column(name = "shipping_postal_code", nullable = false, length = 30)),
            @AttributeOverride(name = "countryCode", column = @Column(name = "shipping_country_code", nullable = false, length = 2))
    })
    private OrderAddressSnapshot shippingAddress;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "label", column = @Column(name = "billing_label", length = 50)),
            @AttributeOverride(name = "recipientName", column = @Column(name = "billing_recipient_name", nullable = false, length = 120)),
            @AttributeOverride(name = "phone", column = @Column(name = "billing_phone", nullable = false, length = 32)),
            @AttributeOverride(name = "line1", column = @Column(name = "billing_line_1", nullable = false, length = 180)),
            @AttributeOverride(name = "line2", column = @Column(name = "billing_line_2", length = 180)),
            @AttributeOverride(name = "city", column = @Column(name = "billing_city", nullable = false, length = 80)),
            @AttributeOverride(name = "state", column = @Column(name = "billing_state", nullable = false, length = 80)),
            @AttributeOverride(name = "postalCode", column = @Column(name = "billing_postal_code", nullable = false, length = 30)),
            @AttributeOverride(name = "countryCode", column = @Column(name = "billing_country_code", nullable = false, length = 2))
    })
    private OrderAddressSnapshot billingAddress;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> orderItems = new ArrayList<>();

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<VendorOrder> vendorOrders = new ArrayList<>();

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    void ensureId() {
        if (id == null) id = UUID.randomUUID();
    }
}

File: microservce-frontend/lib/types/order.ts
export type OrderAddress = {
  sourceAddressId: string;
  label: string | null;
  recipientName: string;
  phone: string;
  line1: string;
  line2: string | null;
  city: string;
  state: string;
  postalCode: string;
  countryCode: string;
};

export type OrderItem = {
  id: string;
  productId: string;
  vendorId: string;
  productSku: string;
  item: string;
  quantity: number;
  unitPrice: number;
  lineTotal: number;
  discountAmount: number;
  fulfilledQuantity: number | null;
  cancelledQuantity: number | null;
};

export type Order = {
  id: string;
  orderNumber?: string;
  customerId: string;
  /** Display-friendly summary string, e.g. "Product A + 2 more". Not an iterable — see OrderDetail.items for the full list. */
  item: string;
  quantity: number;
  itemCount: number;
  orderTotal: number;
  currency: string;
  subtotal: number;
  lineDiscountTotal: number;
  cartDiscountTotal: number;
  shippingAmount: number;
  shippingDiscountTotal: number;
  totalDiscount: number;
  couponCode: string | null;
  couponReservationId: string | null;
  paymentId: string | null;
  paymentMethod: string | null;
  paymentGatewayRef: string | null;
  paidAt: string | null;
  customerNote: string | null;
  status: string;
  expiresAt: string | null;
  refundAmount: number | null;
  refundReason: string | null;
  refundInitiatedAt: string | null;
  refundCompletedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

export type CustomerSummary = {
  id: string;
  name: string;
  email: string;
};

export type OrderDetail = {
  id: string;
  orderNumber?: string;
  customerId: string;
  item: string;
  quantity: number;
  itemCount: number;
  orderTotal: number;
  currency: string;
  subtotal: number;
  lineDiscountTotal: number;
  cartDiscountTotal: number;
  shippingAmount: number;
  shippingDiscountTotal: number;
  totalDiscount: number;
  couponCode: string | null;
  couponReservationId: string | null;
  paymentId: string | null;
  paymentMethod: string | null;
  paymentGatewayRef: string | null;
  paidAt: string | null;
  customerNote: string | null;
  adminNote: string | null;
  status: string;
  expiresAt: string | null;
  refundAmount: number | null;
  refundReason: string | null;
  refundInitiatedAt: string | null;
  refundCompletedAt: string | null;
  createdAt: string;
  updatedAt: string;
  items: OrderItem[];
  shippingAddress: OrderAddress | null;
  billingAddress: OrderAddress | null;
  customer: CustomerSummary | null;
  warnings: string[];
};

export type OrderStatusAudit = {
  id: string;
  fromStatus: string | null;
  toStatus: string;
  actorSub: string;
  actorRoles: string | null;
  actorType: string | null;
  changeSource: string | null;
  note: string | null;
  createdAt: string;
};

export type VendorOrder = {
  id: string;
  orderId: string;
  vendorId: string;
  vendorName?: string;
  status: string;
  itemCount: number;
  quantity: number;
  orderTotal: number;
  currency: string;
  discountAmount: number;
  shippingAmount: number;
  platformFee: number;
  payoutAmount: number;
  trackingNumber: string | null;
  trackingUrl: string | null;
  carrierCode: string | null;
  estimatedDeliveryDate: string | null;
  refundAmount: number | null;
  refundedAmount: number | null;
  refundedQuantity: number | null;
  refundReason: string | null;
  refundInitiatedAt: string | null;
  refundCompletedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

export type CancelOrderRequest = {
  reason?: string;
};

export type InvoiceLineItem = {
  name: string;
  quantity: number;
  unitPrice: number;
  lineTotal: number;
};

export type InvoiceResponse = {
  orderId: string;
  invoiceNumber: string;
  customerName: string;
  customerEmail: string;
  items: InvoiceLineItem[];
  subtotal: number;
  discountTotal: number;
  shippingTotal: number;
  grandTotal: number;
  currency: string;
  issuedAt: string;
};
